File: event-bus/shared/src/main/java/com/github/dynamicextensionsalfresco/event/impl/DefaultEventBus.java
package com.github.dynamicextensionsalfresco.event.impl;

import com.github.dynamicextensionsalfresco.event.Event;
import com.github.dynamicextensionsalfresco.event.EventBus;
import com.github.dynamicextensionsalfresco.event.EventListener;
import java.util.Arrays;
import kotlin.jvm.internal.Intrinsics;
import org.jetbrains.annotations.NotNull;
import org.osgi.framework.BundleContext;
import org.osgi.framework.InvalidSyntaxException;

/**
 * Synchronous implementation that only notifies listeners based on their supported events.
 *
 * @author Laurent Van der Linden
 */
public final class DefaultEventBus implements EventBus {

    private final BundleContext bundleContext;

    public DefaultEventBus(@NotNull BundleContext bundleContext) {
        Intrinsics.checkParameterIsNotNull(bundleContext, "bundleContext");

        this.bundleContext = bundleContext;
    }

    @SuppressWarnings("unchecked")
    public void publish(@NotNull final Event event) {
        Intrinsics.checkParameterIsNotNull(event, "event");

        try {
            Arrays.stream(bundleContext.getAllServiceReferences(EventListener.class.getName(), null))
                    .map(bundleContext::getService)
                    .filter(e -> e instanceof EventListener)
                    .filter(e -> Arrays.stream(((EventListener) e).supportedEventTypes())
                            .anyMatch(it -> it == event.getClass()))
                    .forEach(e -> ((EventListener<Event>) e).onEvent(event));
        } catch (InvalidSyntaxException e) {
            throw new IllegalStateException("Strange, 'InvalidSyntaxException' should not be thrown if filter is null?",
                    e);
        }
    }
}

File: event-bus/shared/src/main/kotlin/com/github/dynamicextensionsalfresco/event/impl/DefaultEventBus.kt
package com.github.dynamicextensionsalfresco.event.impl

import com.github.dynamicextensionsalfresco.event.Event
import com.github.dynamicextensionsalfresco.event.EventBus
import com.github.dynamicextensionsalfresco.event.EventListener
import org.osgi.framework.BundleContext

/**
 * Synchronous implementation that only notifies listeners based on their supported events.

 * @author Laurent Van der Linden
 */
public class DefaultEventBus(private val bundleContext: BundleContext) : EventBus {
    @Suppress("UNCHECKED_CAST")
    override fun publish(event: Event) {
        bundleContext.getAllServiceReferences(EventListener::class.java.name, null)
                .asSequence()
                .map { bundleContext.getService(it) }
                .filterIsInstance(EventListener::class.java)
                .filter { it.supportedEventTypes.any { it == event.javaClass } }
                .forEach { (it as EventListener<Event>).onEvent(event) }
    }
}
